components/messages.js

```diff
@@ -1,54 +1,78 @@
+import { RotateCcw as UndoIcon } from "lucide-react";
 import Image from "next/future/image";
+import { Fragment, useEffect, useRef } from "react";
 import PulseLoader from "react-spinners/PulseLoader";
-import React, { useRef, useEffect } from "react";
+import Message from "./message";
 
-export default function Messages({ events, isProcessing }) {
+export default function Messages({ events, isProcessing, onUndo }) {
   const messagesEndRef = useRef(null);
 
   useEffect(() => {
     if (events.length > 2) {
       messagesEndRef.current.scrollIntoView({ behavior: "smooth" });
     }
-  });
+  }, [events.length]);
 
   return (
-    <div className="canvas w-full">
+    <section className="w-full">
       {events.map((ev, index) => {
         if (ev.image) {
           return (
-            <div key={"event-" + index} className="w-full">
-              <Image
-                alt={"event" + index}
-                width="512"
-                height="512"
-                priority={true}
-                className="w-full h-auto mb-10 rounded-lg"
-                src={ev.image}
-              />
-            </div>
+            <Fragment key={"image-" + index}>
+              <Message sender="replicate" shouldFillWidth>
+                <Image
+                  alt={
+                    ev.prompt
+                      ? `The result of the prompt "${ev.prompt}" on the previous image`
+                      : "The source image"
+                  }
+                  width="512"
+                  height="512"
+                  priority={true}
+                  className="w-full h-auto rounded-lg"
+                  src={ev.image}
+                />
+
+                {onUndo && index > 0 && index === events.length - 1 && (
+                  <div className="mt-2 text-right">
+                    <button
+                      className="lil-button"
+                      onClick={() => {
+                        onUndo(index);
+                      }}
+                    >
+                      <UndoIcon className="icon" /> Undo and try a different
+                      change
+                    </button>
+                  </div>
+                )}
+              </Message>
+
+              {(isProcessing || index < events.length - 1) && (
+                <Message sender="replicate" isSameSender>
+                  What should we change?
```

Comment:
It seems like this only needs to be shown to the user once to get them going, but I'm seeing a second time after generating an image:

![Screenshot 2023-01-29 at 6 41 07 PM](https://user-images.githubusercontent.com/2289/215376384-b088b956-d068-4788-9dc4-ad902fa24981.png)

Reply:
So it's serving the purpose of the label for the input field. Inputs should have visible labels. We could show a different message, though.